src/nix/path-info.cc

```diff
@@ -78,11 +80,12 @@ static json pathInfoToJSON(
 
         } catch (InvalidPath &) {
             jsonObject = nullptr;
+            resultCode = 1;
```

Comment:
Random idea: increment and have exit code be the number of necessary builds, max 255?

Reply:
Usually exit codes refer to the kind of error.
For instance, we could return `1` for arbitrary errors, but return `2` when paths are missing.
This allows scripts to easily and safely recover from missing paths without misidentifying a more "catastrophic" failure such as a `--store` that isn't reachable, which the script is unlikely to be able to recover from.
Also I believe the number of paths is usually irrelevant, because it is a bad proxy for the amount of work to be done (e.g. cheap NixOS configuration rebuilds are easily a dozen, while a kernel rebuild is only one or a few), and scripts would only check whether they need to rebuild anything at all, ie `[[ $r != 0 ]]`.

More generally, if we want to make good use of error codes, we should start by taking inventory of the exit codes we have, documenting them, and then use them consistently. Also I believe they should _not_ be integrated into the exception system, because context matters.
For instance, the missing store error could be `ERR_NOT_FOUND = 44`, and the missing store path exception could use the same code. However the prior should usually be treated as a fatal error, whereas the latter is about the actual CLI operation, which is recoverable.

This problem could be solved by having more specific codes or by making the CLI itself responsible for assigning meaning to the exceptions. I'm leaning towards the latter, because it is not realistic for scripts to anticipate many specific error codes, many of which are irrelevant. I'd favor a solution where we have fewer codes, but they are relevant to the actual commands, and they map to documentation. The global option flags situation should serve as a warning.

Reply:
I've split this out into #11099

Reply:
While I agree with what you've said @roberth that is a much larger change.

My thoughts:
We could still merge this as it fixes the bug and cements the behavior of what to do in the subsequent larger change for exit codes.
or
leave it but at least document with a `TODO` or something?

Additionally:
Confirming: JSON exit always with -1 ?
You could test it from the value of the JSON. I want to also make sure that this wasn't the intent to have JSON always return 0 and the status code be implicit from the object being returned.

Reply:
I think returning `1` is already an improvement, and we can make it more specific later. A different exit code behavior for `--json` doesn't seem right to me.
Also, making it a more specific code would create an impression that we've already picked a numbering system, so `1` is the best we can do before designing that.

> -1

Negative exit codes (or equivalently, >127) are assumed to reflect signal codes.
```
$ ./segfault 
Segmentation fault (core dumped)

$ echo $?
139
```

Reply:
Sounds good.